From cfea53fc2bf7095273c6731dbbc35cdf5445d38c Mon Sep 17 00:00:00 2001
From: "Christopher A. Williamson" <contact@chrisaw.io>
Date: Sun, 1 Jun 2025 14:45:04 +0100
Subject: [PATCH 3/17] Allow EXPENSIVE_RENDERING mode to be disabled

Suppress the EXPENSIVE_RENDERING power mode hint in PowerManagerService
when persist.sys.phh.disable_expensive_rendering_mode is set. Fixes
constant log spam on some MediaTek devices.

Companion patch: staging/platform_frameworks_native/0001 (SurfaceFlinger
side of the same feature).
---
 .../jni/com_android_server_power_PowerManagerService.cpp    | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/services/core/jni/com_android_server_power_PowerManagerService.cpp b/services/core/jni/com_android_server_power_PowerManagerService.cpp
index a11b40ea4..a5b8739ed 100644
--- a/services/core/jni/com_android_server_power_PowerManagerService.cpp
+++ b/services/core/jni/com_android_server_power_PowerManagerService.cpp
@@ -42,6 +42,7 @@
 #include <utils/String8.h>
 #include <utils/Timers.h>
 #include <utils/misc.h>
+#include <android-base/properties.h>

 #include "jni.h"

@@ -84,6 +85,11 @@ static void setPowerBoost(Boost boost, int32_t durationMs) {
 }

 static bool setPowerMode(Mode mode, bool enabled) {
+    // disable EXPENSIVE_RENDERING mode
+    if (android::base::GetBoolProperty("persist.sys.phh.disable_expensive_rendering_mode", false) && mode == Mode::EXPENSIVE_RENDERING) {
+        return true; // pretend the operation succeeded
+    }
+
     android::base::Timer t;
     auto result = gPowerHalController.setMode(mode, enabled);
     if (mode == Mode::INTERACTIVE && t.duration() > 20ms) {
-- 
2.53.0

